cfox/utils/reswrap.1
.TH reswrap 1 "February 2022" "The FOX C++ GUI Toolkit" "FOX Programmers Manual"

.SH NAME
reswrap \- wrap icon and image resources into C or C++ code
.SH SYNOPSIS
.B reswrap
.RB [ options ]
.RB [ \-o [ a "] "
.IR outfile ]
.I files...
.SH DESCRIPTION
.B Reswrap
is a tool to turn binary files into C or C++ data arrays.
This allows various resources such as icons, images, or other data to be embedded into source
code and compiled into an application. When used in text mode, normal printing characters
are passed normally while control characters or non-ascii are automatically escaped.

In development of graphical applications with FOX, it is customary to embed icons, images,
online help, or OpenGL fragment shader programs into the executable file, thus removing the
need to distribute such files separately with the application.

.B Reswrap
is typically invoked as a part of the application build process to generate additional C++ source
files from collections of icons, images, or other binary resources. This can be easily accomplished
using a special Makefile Rule or Custom Build Rule, as part of the normal compile process of
an application.
Reswrap supports a variety of options to facilitate inclusion into an automated build process.
Given a list of resource files, reswrap can automatically generate header files containing just
the declarations, adorn the generated symbols with namespaces, prefixes and/or postfixes,
file-extensions, and apply extern or static linkage attributes.  Resource filenames containing
unsafe characters are automatically transformed into legal C or C++ identifiers.

By default reswrap will interpret the
.I files
listed as a stream of raw bytes and output them as an initialized data array to stdout.

.SH OPTIONS
.TP
\fB\-?\fR, \fB\--help\fR
Print summary usage information of all the supported options.

.TP
\fB\-v\fR, \fB\--version\fR
Print the version number and license information.

.TP
\fB\-h\fR, \fB\--header\fR
Generate an output file containing only declarations suitable for including as a header
file.  This also enables the \fI --extern\fR option automatically.

.TP
\fB\-s\fR, \fB\--source\fR
Generate an output file containing data arrays.  This option disables the \fI --extern\fR option.
This option is the default when no directives are given.

.TP
\fB\-V\fR, \fB\--verbose\fR
Print a list of the filenames, and the corresponding resource declarations generated as reswrap
processes the resource files.

.TP
\fB\-i\fR\fI file\fR, \fB\--include\fR\fI file\fR
Generate a \fB#include "\fR\fIfile\fR\fB"\fR directive in the output file, to make the declarations
available when compiling the data source file.

.TP
\fB\-o\fR\fI file\fR, \fB\--output\fR\fI file\fR
Write the output of the conversion into
.I file
instead of writing to stdout.

.TP
\fB\-oa\fR\fI file\fR, \fB\--append\fR\fI file\fR
Append the output of the conversion to
.I file
instead of writing to stdout.  If \fI file\fR does not exist yet, it is created, and the optional
comment line will be written, followed by the optional \fB#include "\fR\fIfile\fR\fB"\fR directive.
If the file already exists then it will be opened for appending.

.TP
\fB\-e\fR, \fB\--extern\fR
Places the storage modifier
.I extern
in front of the data array, ensuring that the data array can be linked with other compilation units.
Normally, constant declarations are not visible in other compilation units.  This option is automatically
turned on when \fI\--header\fR is passed.

.TP
\fB\-S\fR, \fB\--static\fR
Places the storage modifier
.I static
in front of the data array.  This makes the symbols invisible outside the compilation unit.

.TP
\fB\-z\fR, \fB\--size\fR
Output the size of the resource in the declaration of a resource-array.
This allows the \fBsizeof()\fR operator to return the correct size of the resource
even for external declarations.
Note that in text mode, (\fI--text\fR or \fI--ascii\fR options), an extra byte is added to the
size for the end of string character.

.TP
\fB\-d\fR, \fB\--decimal\fR
Write data as decimal numbers instead of using the default hexadecimal numbers.

.TP
\fB\-x\fR, \fB\--hex\fR
Write data as hexadecimal numbers. This option is the default.

.TP
\fB\-t\fR, \fB\--text\fR
Write data as a text string, with each byte represented as a hexadecimal
excape sequence, as in "\\x33".
The C++ compiler appends a nul-character at the end of the text string, thus
making the data array one character longer than the resource file.

.TP
\fB\-a\fR, \fB\--ascii\fR
Write data as a text string, with each byte printed normally unless it must be
escaped, as in "\\x33".  Use this option if you need to embed large chunks of
text (for example, on-line help) into the application.  This lets you keep
the original text in a plain text file, and obviates the need to manually escape
special characters to embed them into source code.
The C++ compiler appends a nul-character at the end of the text string, thus
making the data array one character longer than the resource file.

With the \fI--ascii\fR option, printable ascii characters are passed unescaped,
while special characters like tabs and newlines are given the usual escape codes.

.TP
\fB\-k\fR, \fB\--keep-ext\fR
This option causes reswrap to keep the file extension, replacing the "."
with an underscore "_".
For example, \fIimage.gif\fR generates \fIconst unsigned char image_gif[]\fR.

This option is recommended as it reduces errors when using the data arrays.

.TP
\fB\-nk\fR, \fB\--drop-ext\fR
This option causes the declaration to be based only on the basename of
the resource file.  For example, \fIimage.gif\fR generates \fIconst unsigned char image[]\fR.

.TP
\fB\-m\fR, \fB\--msdos\fR
Read files with MS-DOS mode.  This replaces "\\r\\n" with "\\n" when reading the resource file.
Only use this if the resource is actually text!

.TP
\fB\-b\fR, \fB\--binary\fR
Read files in BINARY mode. Data other than text should always be read in using binary mode.
This option is the default.

.TP
\fB\-u\fR, \fB\--unsigned\fR
Generate \fIunsigned char\fR declaration, even when  \fI--text\fR or \fI--ascii\fR
was specified.

.TP
\fB\-N\fR, \fB\--no-const\fR
Do not generate \fIconst\fR declarations.

.TP
\fB\-C\fR, \fB\--const\fR
Force generation of \fIconst\fR declarations.

.TP
\fB\-cc\fR, \fB\--comments\fR
Add comments to the output files.  This option is the default.  At the top of
each output file, write the current date, and the reswrap version number.

.TP
\fB\-nc\fR, \fB\--no-comments\fR
Suppress generations of comments in the output files.

.TP
\fB\-p\fI name\fR, \fB\--prefix\fI name\fR
Prepend the given
.I prefix
in front of the name of the resource; this may be used to generate class
names or namespace names in front of symbols.

.TP
\fB\-f\fI name\fR, \fB\--suffix\fI name\fR
Prepend the given
.I prefix
in front of the name of the resource; this may be used to generate class
names or namespace names in front of symbols.

.TP
\fB\-n\fI name\fR, \fB\--namespace\fI name\fR
Generate all declarations inside the given C++
.I namespace
declaration.
Using a
.I namespace
may be used to ensure that declarations are only accessible within the
given scope, and thus won't clash with symbols.

.TP
\fB\-c\fI cols\fR, \fB\--columns\fI cols\fR
Writes
.I cols
columns instead of the default number of columns in the data statements
generated by reswrap.
The default number of columns for decimal is 10 columns, while the default for hexadecimal printout is 16 columns.
For text string printout, the default is 80 characters.
When the number of columns is set to 0, decimal or hexadecimal output is printed without breaking
columns, while text output lines are broken only at the newlines.

.TP
\fB\-r\fI name\fR, \fB\--resource\fI name\fR
Instead of using a resource name based on the filename, reswrap substitutes
.I name
for the resource name used in the declaration or definition for the following
resource file.  This is useful if the filename can not be used as an identifier,
for example if the filename happens to be a reserved word in C or C++, like "while.gif".

.SH CAVEATS
When using the text string mode, please remember the C compiler appends
one nul-character after the string, making the data array one element
longer than the resource file.
.SH AUTHOR
This manpage was originally written by Torsten Landschoff (torsten@debian.org)
for the Debian distribution of the FOX Toolkit, and is currently maintained
by Jeroen van der Zijp (jeroen@fox-toolkit.net).
.SH COPYRIGHT
Copyright \(co 1997,2022 Jeroen van der Zijp.
.br
This is free software; see the source for copying conditions.  There is
NO warranty; not even for MERCHANTABILITY or FITNESS FOR  A  PARTICULAR
PURPOSE.
.SH SEE ALSO
The reswrap program is a part of the FOX GUI Toolkit.  Further information
about reswrap can be found at:

.B http://www.fox-toolkit.org

The FOX Toolkit website includes detailed information about reswrap and its use.



ChangeLog
1.2.2
  - gap: added sndio output plugin
  - gap: added crossfader
  - ui: Updated FOX to 1.7.74
  - ui: Updated German translation thanks to Martin Wagner
  - ui: New Korean translation by JungHee Lee 
  - ui: make it easier to adjust cover size in the miniplayer for hdpi screens

1.2.1
  - Update to latest appstream spec.
  - Fix play/pause issue #83

1.2.0
  - Support for non-16bit alac files.
  - Support for aacPlus streams.
  - A copy of FOX is now included to make it easier to distribute gogglesmm
  - Preserve clipboard (copy/paste) when closing track editor.
  - Updated spanish translation
  - Fix buffer overflow in audioscrobbler (issue 77).
  - Fix prevent consecutive spaces when renaming audio files.

1.1.9
  - Fix aac fatal decoder error
  - Fix llvm build with latest sqlite library
  - Fix unused openssl 1.1 api calls
  - Fix various newly gcc7 warnings

1.1.8
  - Fix: resize (if needed) icons from external icon themes on start up
  - Fix: set default sort column for podcasts to date.
  - Fix: typo in ifdef

1.1.7
  - Sort albums by album-artists when displaying covers
  - Add commandline option to refresh podcast feeds
  - Add the ability to build with static FOX library
  - Fix covercache loading with FOX-1.7.57
  - Regenerate covercache if supported file format has changed in FOX library.
  - gap: refactor input plugins using new iocontext
  - gap: remove unimplemented smb plugin

1.1.6
  - https support with OpenSSL or GnuTLS.
  - http(s) gzip support.
  - Local Source: only list playback supported formats
  - New import option to exclude any files that cannot be played back.
  - Moved exclude import filters to its own tab.
  - cmake: added more options to enable/disable plugins during compilation
  - gap: refactor decoder and reader plugins
  - Fix: refresh path lut if needed.
  - Fix: reset scroll position when loading new lyrics
  - Fix: show audio device options when using mousewheel or keys to change driver
  - Fix: tiptext for album list items in case when information is missing
  - Fix: incorrect access check in socket.
  - Fix: volume icon in miniplayer didn't get updated on icon theme change
  - Fix: init directory in open dialog for Play File or Stream

1.1.5
  - Fix lyrics column not being created on first run.
  - Fix import writing lyrics to the wrong column.

1.1.4
  - Lyrics support
  - Matroska playback (only) support
  - Refactor Vorbis / Opus / AAC decoders to support matroska
  - Optimizations to Ogg based decoders should use less memory
  - Refactor text converter and renamer
  - Refactor socket io for improved win32 support
  - FOX 1.7.57 compatibility
  - fix: minor changes to api for text compose/decompose which broke filename formatter.
  - fix: Renaming filename preview dialog now correctly converts non-Ascii/UTF-8 back to UTF-8.
  - fix: Don't substitute & for && when displaying track in title bar

1.1.3
  - Fix deadlock in get_decoder_packet and replace pop_if/wait with wait_for
  - Fix missing config option for tremor.

1.1.2
  - Fix missing option No Resample in alsa output plugin

1.1.1
  - Fix detection of newer kde desktop sessions.
  - Replace gogglesmm native dbus interface with mpris v2.
  - Implement missing functionality in mpris v2.
  - gap: fix file input to handle file:// uri scheme
  - rss: check for updated cover art image
  - http: force mimetype to be lower case.
  - http: reset dir on IO after reading last header in chunked response.
  - Remove remaining lirc functionality.
  - Remove (broken) AppStatus notification implementation.
  - Remove unused settings.

1.1.0
  - flac: use new reader to parse flac files.
  - gap: use more portable signal and semaphore primitives in engine.
  - gap: refactor input_plugin.
  - gap: add proper interface class to OutputThread.
  - gap: simplify includes.
  - gap: fix build issue with mp4/m4a support.
  - remove old 'configure' based buildsystem.
  - lirc: remove unused/unsupported lirc support
  - c++11: NULL -> nullptr.
  - replace glew with epoxy.
  - fix crash editing radio station. (issue 56)

1.0.18
  - ALAC codec support.
  - add hashbang to installed python scripts
  - fix vorbis comment parsing
  - fix opus filetype detection during import.
  - remove debug statements during release builds.
  - cmake: fix reverse check for DESTDIR
  - cmake: build mp4 reader when faad is enabled

1.0.17
  - cmake: fix libgcrypt detection and link to internal md5 if no gcrypt is available

1.0.16
  - cmake: fix gogglesmm not using BUILD_GAP_SHARED_LIB variable

1.0.15

cfox/utils/reswrap.cpp
/********************************************************************************
*                                                                               *
*                R e s o u r c e   W r a p p i n g   U t i l i t y              *
*                                                                               *
*********************************************************************************
* Copyright (C) 1997,2022 by Jeroen van der Zijp.   All Rights Reserved.        *
*********************************************************************************
* This program is free software: you can redistribute it and/or modify          *
* it under the terms of the GNU General Public License as published by          *
* the Free Software Foundation, either version 3 of the License, or             *
* (at your option) any later version.                                           *
*                                                                               *
* This program is distributed in the hope that it will be useful,               *
* but WITHOUT ANY WARRANTY; without even the implied warranty of                *
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                 *
* GNU General Public License for more details.                                  *
*                                                                               *
* You should have received a copy of the GNU General Public License             *
* along with this program.  If not, see <http://www.gnu.org/licenses/>.         *
********************************************************************************/
#include "stdio.h"
#include "stdlib.h"
#include "string.h"
#include <time.h>
#include "ctype.h"



/*

  Notes:
  - Can now also generate output as a (possibly escaped) text string.
  - Options to suffixes and prefixes; for example:

        reswrap -o icons.cpp  --prefix tb_ --suffix _gif *.gif --suffix _bmp *.bmp

    Given file names a.gif, b.gif, c.bmp, d.bmp, generates names like:

        const unsigned char tb_a_gif[];
        const unsigned char tb_b_gif[];
        const unsigned char tb_c_bmp[];
        const unsigned char tb_d_bmp[];

  - Added ability to switch output files allows both icons.h and icons.cpp
    to be generated in one command like:

        reswrap -h -o icons.h $(ICONS) -s -i icons.h -o icons.cpp $(ICONS)

    which will kill issues with dependencies; this will also embed the
    string:

        #include "icons.h"

    Into the icons.cpp file.
*/

#define MODE_DECIMAL   0
#define MODE_HEX       1
#define MODE_TEXT      2
#define MODE_ASCII     3

#define LINKAGE_NONE   0
#define LINKAGE_EXTERN 1
#define LINKAGE_STATIC 2

#define TYPE_SOURCE    0
#define TYPE_HEADER    1

#define MAX_RESOURCE   512

#define NUM_COLUMNS    16

/*******************************************************************************/


const char version[]="6.0.0";


typedef struct {
  const char *outfilename;
  const char *prefix;
  const char *suffix;
  const char *scope;
  const char *header;
  FILE       *outfile;
  int         append;
  int         maxcols;
  int         colsset;
  int         linkage;
  int         filetype;
  int         declaresize;
  int         forceunsigned;
  int         constant;
  int         comments;
  int         keepext;
  int         binary;
  int         verbose;
  int         mode;
  } OPTIONS;

/*******************************************************************************/

/* Print short help */
static void printhelp(const char *option){
  printf("reswrap: invalid option: -- %s\n",option);
  printf("Usage: reswrap [options] [-o[a] outfile] files...\n");
  printf("Try \"reswrap --help\" for more information.\n");
  }


/* Print some help */
static void printusage(){
  printf("Usage: reswrap [options] [-o[a] outfile] files...\n");
  printf("Convert files containing images, text, or binary data into C/C++ data arrays.\n");
  printf("\n");
  printf("Options:\n");
  printf("  -?, --help                Print this help\n");
  printf("  -v, --version             Print version number\n");
  printf("  -h, --header              Create header file containing only declarations\n");
  printf("  -s, --source              Create source file containing data arrays (default)\n");
  printf("  -V, --verbose             Show which resource files are being processed\n");
  printf("  -i file, --include file   Generate #include \"file\" in output file\n");
  printf("  -o file, --output file    Output to file instead of stdout\n");
  printf("  -oa file, --append file   Append to file instead of stdout\n");
  printf("  -e, --extern              Generate extern reference declarations (default for headers)\n");
  printf("  -S, --static              Generate static reference declarations\n");
  printf("  -z, --size                Output size in array declarations\n");
  printf("  -d, --decimal             Output as decimal\n");
  printf("  -x, --hex                 Output as hex (default)\n");
  printf("  -t, --text                Output as hexadecimal text string\n");
  printf("  -a, --ascii               Output as ascii text string\n");
  printf("  -k, --keep-ext            Keep file extension, replacing period by underscore\n");
  printf("  -nk, --drop-ext           Drop extension (default)\n");
  printf("  -m, --msdos               Read files with MS-DOS mode\n");
  printf("  -b, --binary              Read files using binary mode (default)\n");
  printf("  -u, --unsigned            Force unsigned char even for text mode\n");
  printf("  -N  --no-const            Don't output const specifier in declaration\n");
  printf("  -C  --const               Force const specifier in declaration\n");
  printf("  -cc, --comments           Add comments to the output files (default)\n");
  printf("  -nc, --no-comments        Remove comments from the output files\n");
  printf("  -p name, --prefix name    Prepend name in front of names of declarations and definitions\n");
  printf("  -f name, --suffix name    Append name in after names of declarations and definitions\n");
  printf("  -n name, --namespace name Place declarations and definitions inside namespace name\n");
  printf("  -c cols, --columns cols   Change number of columns in output from %d to cols\n",NUM_COLUMNS);
  printf("  -r name, --resource name  Override resource name of following resource file\n");
  printf("\n");
  }


/* Print version information */
static void printversion(){
  printf("reswrap %s\n\n",version);
  printf("Copyright (C) 1997,2022 Jeroen van der Zijp. All Rights Reserved.\n");
  printf("Please visit: http://www.fox-toolkit.org for further information.\n");
  printf("\n");
  printf("This program is free software: you can redistribute it and/or modify\n");
  printf("it under the terms of the GNU General Public License as published by\n");
  printf("the Free Software Foundation, either version 3 of the License, or\n");
  printf("(at your option) any later version.\n");
  printf("\n");
  printf("This program is distributed in the hope that it will be useful,\n");
  printf("but WITHOUT ANY WARRANTY; without even the implied warranty of\n");
  printf("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n");
  printf("GNU General Public License for more details.\n");
  printf("\n");
  printf("You should have received a copy of the GNU General Public License\n");
  printf("along with this program.  If not, see <http://www.gnu.org/licenses/>.\n");
  }

/*******************************************************************************/

/* Build resource name */
static const char* resourcename(char *name,const char* filename,int keepdot){
  const char* begin=name;
  const char* ptr;

  /* Get name only; take care of mixed path separator characters on mswindows */
  if((ptr=strrchr(filename,'/'))!=0) filename=ptr;
  if((ptr=strrchr(filename,'\\'))!=0) filename=ptr;

  /* Copy filename */
  while(*filename){

    /* C++ identifier may contain _, alpha, digit (if not first), or namespace separator : */
    if(*filename==':' || *filename=='_' || isalpha(*filename) || (isdigit(*filename) && (begin!=name))){
      *name++=*filename;
      }

    /* We can squash dot extension to _ */
    else if(*filename=='.'){
      if(!keepdot) break;
      *name++='_';
      }

    filename++;
    }
  *name=0;
  return begin;
  }


/* Process prologue */
static void prologue(OPTIONS* opts){
  char date[100];
  time_t clock;
  if(opts->outfilename){
    if(opts->append){
      opts->outfile=fopen(opts->outfilename,"a");
      if(!opts->outfile){
        fprintf(stderr,"reswrap: unable to open output file: %s\n",opts->outfilename);
        exit(1);
        }
      if(ftell(opts->outfile)>0) return;
      }
    else{
      opts->outfile=fopen(opts->outfilename,"w");
      if(!opts->outfile){
        fprintf(stderr,"reswrap: unable to open output file: %s\n",opts->outfilename);
        exit(1);
        }
      }
    }
  if(opts->comments){
    clock=time(NULL);
    strftime(date,sizeof(date),"%Y/%m/%d %H:%M:%S",localtime(&clock));
    fprintf(opts->outfile,"/*********** Generated on %s by reswrap version %s *********/\n\n",date,version);
    }
  if(opts->header){
    fprintf(opts->outfile,"#include \"%s\"\n\n",opts->header);
    }
  if(opts->scope){
    fprintf(opts->outfile,"namespace %s {\n\n",opts->scope);
    }
  }


/* Process epilogue */
static void epilogue(OPTIONS* opts){
  if(opts->scope){
    fprintf(opts->outfile,"}\n");
    }
  if(opts->outfile!=stdout){
    fclose(opts->outfile);
    opts->outfile=stdout;
    }
  }


/* Process resource file */
static int processresourcefile(const char* filename,const char* name,OPTIONS* opts){
  const char *typestr="unsigned char ";
  const char *sizestr="[]";
  const char *conststr="";
  const char *linkstr="";
  char resource[MAX_RESOURCE],size[10];
  int ressize,first,col,hex,b;
  FILE *file;

  /* Open resource file; always open as binary */
  file=fopen(filename,"rb");
  if(file){

    /* Get the size */
    fseek(file,0,SEEK_END);
    ressize=ftell(file);
    fseek(file,0,SEEK_SET);

    /* Add one if text mode, for end of string */
    if(opts->mode>=MODE_TEXT){
      ressize++;
      }

    /* Output header */
    if(opts->comments){
      fprintf(opts->outfile,"/* Created by reswrap from file %s */\n",filename);
      }

    /* Generate external reference */
    if(opts->linkage==LINKAGE_EXTERN){
      linkstr="extern ";
      }

    /* Generate static reference */
    else if(opts->linkage==LINKAGE_STATIC){
      linkstr="static ";
      }

    /* Generate const declaration */
    if(opts->constant){
      conststr="const ";
      }

    /* In text mode, output a 'char' declaration */
    if((opts->mode>=MODE_TEXT) && !opts->forceunsigned){
      typestr="char ";
      }

    /* Compute resource name from filename if no override */
    if(!name){
      name=resourcename(resource,filename,opts->keepext);
      }

    /* Size specifier */
    if(opts->declaresize){
      sprintf(size,"[%d]",ressize);
      sizestr=size;
      }

    /* Generate resource name */
    fprintf(opts->outfile,"%s%s%s%s%s%s%s",linkstr,conststr,typestr,opts->prefix,name,opts->suffix,sizestr);

    /* Generating source file */
    if(opts->filetype==TYPE_SOURCE){

      /* Text Mode */
      if(opts->mode>=MODE_TEXT){
        col=0;
        hex=0;
        fprintf(opts->outfile,"=\n  \"");
        while((b=fgetc(file))!=EOF){
          if(!opts->binary && (b=='\r')) continue;
          if(col>=opts->maxcols && opts->maxcols){
            fprintf(opts->outfile,"\"\n  \"");
            col=0;
            }
          if(opts->mode==MODE_ASCII){
            if(b=='\\'){ fprintf(opts->outfile,"\\\\"); col+=2; hex=0; }
            else if(b=='\a'){ fprintf(opts->outfile,"\\a"); col+=2; hex=0; }
            else if(b=='\t'){ fprintf(opts->outfile,"\\t"); col+=2; hex=0; }
            else if(b=='\r'){ fprintf(opts->outfile,"\\r"); col+=2; hex=0; }
            else if(b=='\f'){ fprintf(opts->outfile,"\\f"); col+=2; hex=0; }
            else if(b=='\v'){ fprintf(opts->outfile,"\\v"); col+=2; hex=0; }
            else if(b=='\"'){ fprintf(opts->outfile,"\\\""); col+=2; hex=0; }
            else if(b=='\n'){ fprintf(opts->outfile,"\\n\"\n  \""); col=0; hex=0; }
            else if(b<32 || b>=127){ fprintf(opts->outfile,"\\x%02x",b); col+=4; hex=1; }
            else if(hex && isxdigit(b)){ fprintf(opts->outfile,"\\x%02x",b); col+=4; hex=1; }
            else{ fprintf(opts->outfile,"%c",b); col+=1; hex=0; }
            }
          else{
            fprintf(opts->outfile,"\\x%02x",b); col+=4;
            }
          }
        fprintf(opts->outfile,"\"\n  ");
        }

      /* Normal Mode */
      else{
        col=0;
        first=1;
        fprintf(opts->outfile,"={");
        if(opts->maxcols) fprintf(opts->outfile,"\n  ");
        while((b=fgetc(file))!=EOF){
          if(!opts->binary && (b=='\r')) continue;
          if(!first){
            fprintf(opts->outfile,",");
            }
          if(col>=opts->maxcols && opts->maxcols){
            fprintf(opts->outfile,"\n  ");
            col=0;
            }
          if(opts->mode==MODE_HEX){
            fprintf(opts->outfile,"0x%02x",b);
            }
          else{
            fprintf(opts->outfile,"%3d",b);
            }
          first=0;
          col++;
          }
        if(opts->maxcols) fprintf(opts->outfile,"\n  ");
        fprintf(opts->outfile,"}");
        }
      }

    /* Append ; */
    fprintf(opts->outfile,";\n\n");

    /* Verbose mode */
    if(opts->verbose){
      fprintf(stderr,"%-30s  %s%s%s%s%s%s%s\n",filename,linkstr,conststr,typestr,opts->prefix,name,opts->suffix,sizestr);
      }

    /* Close resource file */
    fclose(file);
    return 1;
    }
  return 0;
  }

/*******************************************************************************/

/* Main */
int main(int argc,char **argv){
  const char *resource=0;
  int         needprologue=1;
  int         needepilogue=0;
  OPTIONS     opts;
  int         arg;

  /* Initialize */
  opts.outfilename=0;
  opts.prefix="";
  opts.suffix="";
  opts.scope=0;
  opts.append=0;
  opts.header=0;
  opts.constant=1;
  opts.outfile=stdout;
  opts.maxcols=NUM_COLUMNS;
  opts.colsset=0;
  opts.linkage=LINKAGE_NONE;
  opts.filetype=TYPE_SOURCE;
  opts.declaresize=0;
  opts.forceunsigned=0;
  opts.comments=1;
  opts.keepext=0;
  opts.binary=1;
  opts.verbose=0;
  opts.mode=MODE_HEX;

  /* Process all options first, except for the -r option */
  for(arg=1; arg<argc; arg++){

    /* Handle options */
    if(argv[arg][0]=='-'){

      /* Change output file */
      if(strcmp(argv[arg],"-o")==0 || strcmp(argv[arg],"--output")==0){

        /* Output epilogue */
        if(needepilogue){
          epilogue(&opts);
          }

        /* Check if argument provided */
        if(++arg>=argc){
          fprintf(stderr,"reswrap: missing argument for -o or --output option\n");
          exit(1);
          }

        /* Get new filename */
        opts.outfilename=argv[arg];
        opts.append=0;
        needprologue=1;
        needepilogue=0;
        }

      /* Change output file and append */
      else if(strcmp(argv[arg],"-oa")==0 || strcmp(argv[arg],"--append")==0){

        /* Output epilogue */
        if(needepilogue){
          epilogue(&opts);
          }

        /* Check if argument provided */
        if(++arg>=argc){
          fprintf(stderr,"reswrap: missing argument for -oa or --append option\n");
          exit(1);
          }

        /* Get new filename */
        opts.outfilename=argv[arg];
        opts.append=1;
        needprologue=1;
        needepilogue=1;
        }

      /* Print help */
      else if(strcmp(argv[arg],"-?")==0 || strcmp(argv[arg],"--help")==0 || (strcmp(argv[arg],"-h")==0 && arg==argc-1)){
        printusage();
        exit(0);
        }

      /* Building include file */
      else if(strcmp(argv[arg],"-h")==0 || strcmp(argv[arg],"--header")==0){
        opts.filetype=TYPE_HEADER;
        opts.linkage=LINKAGE_EXTERN;
        }

      /* Building source file */
      else if(strcmp(argv[arg],"-s")==0 || strcmp(argv[arg],"--source")==0){
        opts.filetype=TYPE_SOURCE;
        opts.linkage=LINKAGE_NONE;
        }

      /* Include header file */
      else if(strcmp(argv[arg],"-i")==0 || strcmp(argv[arg],"--include")==0){
        if(++arg>=argc){
          fprintf(stderr,"reswrap: missing filename for -i or --include option\n");
          exit(1);
          }
        opts.header=0;
        if(strcmp(argv[arg],"-")!=0){
          opts.header=argv[arg];
          }
        }

      /* Print version */
      else if(strcmp(argv[arg],"--version")==0){
        printversion();
        exit(0);
        }

      /* Print version */
      else if(strcmp(argv[arg],"-v")==0){
        printf("reswrap %s\n",version);
        exit(0);
        }

      /* Switch to decimal */
      else if(strcmp(argv[arg],"-d")==0 || strcmp(argv[arg],"--decimal")==0){
        opts.mode=MODE_DECIMAL;
        if(!opts.colsset) opts.maxcols=10;
        }

      /* Switch to hex */
      else if(strcmp(argv[arg],"-x")==0 || strcmp(argv[arg],"--hex")==0){
        opts.mode=MODE_HEX;
        if(!opts.colsset) opts.maxcols=16;
        }

      /* Switch to hexadecimal text */
      else if(strcmp(argv[arg],"-t")==0 || strcmp(argv[arg],"--text")==0){
        opts.mode=MODE_TEXT;
        if(!opts.colsset) opts.maxcols=80;
        }

      /* Switch to ascii text */
      else if(strcmp(argv[arg],"-a")==0 || strcmp(argv[arg],"--ascii")==0){
        opts.mode=MODE_ASCII;
        if(!opts.colsset) opts.maxcols=80;
        }

      /* Switch to ascii text */
      else if(strcmp(argv[arg],"-ta")==0){
        fprintf(stderr,"reswrap: -ta option is deprecated; please use -a or --ascii\n");
        opts.mode=MODE_ASCII;
        if(!opts.colsset) opts.maxcols=80;
        }

      /* Generate as external reference */
      else if(strcmp(argv[arg],"-e")==0 || strcmp(argv[arg],"--extern")==0){
        opts.linkage=LINKAGE_EXTERN;
        }

      /* Generate as static reference */
      else if(strcmp(argv[arg],"-S")==0 || strcmp(argv[arg],"--static")==0){
        opts.linkage=LINKAGE_STATIC;
        }

      /* Force unsigned */
      else if(strcmp(argv[arg],"-u")==0 || strcmp(argv[arg],"--unsigned")==0){
        opts.forceunsigned=1;
        }

      /* No constant declaration */
      else if(strcmp(argv[arg],"-N")==0 || strcmp(argv[arg],"--no-const")==0){
        opts.constant=0;
        }

      /* Force constant declaration */
      else if(strcmp(argv[arg],"-C")==0 || strcmp(argv[arg],"--const")==0){
        opts.constant=1;
        }

      /* Declare size */
      else if(strcmp(argv[arg],"-z")==0 || strcmp(argv[arg],"--size")==0){
        opts.declaresize=1;
        }

      /* Read resource with MS-DOS mode */
      else if(strcmp(argv[arg],"-m")==0 || strcmp(argv[arg],"--msdos")==0){
        opts.binary=0;
        }

      /* Read resource with BINARY mode */
      else if(strcmp(argv[arg],"-b")==0 || strcmp(argv[arg],"--binary")==0){
        opts.binary=1;
        }

      /* Keep extension */
      else if(strcmp(argv[arg],"-k")==0 || strcmp(argv[arg],"--keep-ext")==0){
        opts.keepext=1;
        }

      /* Base name only */
      else if(strcmp(argv[arg],"-nk")==0 || strcmp(argv[arg],"--drop-ext")==0){
        opts.keepext=0;
        }

      /* Change number of columns */
      else if(strcmp(argv[arg],"-c")==0 || strcmp(argv[arg],"--columns")==0){
        if(++arg>=argc){
          fprintf(stderr,"reswrap: missing argument for -c or --columns option\n");
          exit(1);
          }
        if(sscanf(argv[arg],"%d",&opts.maxcols)==1 && opts.maxcols<0){
          fprintf(stderr,"reswrap: illegal argument for number of columns\n");
          exit(1);
          }
        opts.colsset=1;
        }

      /* Add comments */
      else if(strcmp(argv[arg],"-cc")==0 || strcmp(argv[arg],"--comments")==0){
        opts.comments=1;
        }

      /* No comments */
      else if(strcmp(argv[arg],"-nc")==0 || strcmp(argv[arg],"--no-comments")==0){
        opts.comments=0;
        }

      /* Verbose */
      else if(strcmp(argv[arg],"-V")==0 || strcmp(argv[arg],"--verbose")==0){
        opts.verbose=1;
        }

      /* Embed in namespace */
      else if(strcmp(argv[arg],"-n")==0 || strcmp(argv[arg],"--namespace")==0){
        if(++arg>=argc){
          fprintf(stderr,"reswrap: missing argument for -n option\n");
          exit(1);
          }
        opts.scope=0;
        if(strcmp(argv[arg],"-")!=0){
          opts.scope=argv[arg];
          }
        }

      /* Prefix in front of declarations */
      else if(strcmp(argv[arg],"-p")==0 || strcmp(argv[arg],"--prefix")==0){
        if(++arg>=argc){
          fprintf(stderr,"reswrap: missing argument for -p or --prefix option\n");
          exit(1);
          }
        opts.prefix="";
        if(strcmp(argv[arg],"-")!=0){
          opts.prefix=argv[arg];
          }
        }

      /* Suffix behind declarations */
      else if(strcmp(argv[arg],"-f")==0 || strcmp(argv[arg],"--suffix")==0){
        if(++arg>=argc){
          fprintf(stderr,"reswrap: missing argument for -f or --suffix option\n");
          exit(1);
          }
        opts.suffix="";
        if(strcmp(argv[arg],"-")!=0){
          opts.suffix=argv[arg];
          }
        }

      /* Explicitly named resource for this filename */
      else if(strcmp(argv[arg],"-r")==0 || strcmp(argv[arg],"--resource")==0){
        if(++arg>=argc){
          fprintf(stderr,"reswrap: missing argument for -r or --resource option\n");
          exit(1);
          }
        resource=argv[arg];
        }

      /* Don't know this option */
      else{
        printhelp(argv[arg]);
        exit(1);
        }
      }

    /* Handle resource file */
    else{

      /* Output prologue */
      if(needprologue){
        prologue(&opts);
        needprologue=0;
        }

      /* Process resource file */
      if(!processresourcefile(argv[arg],resource,&opts)){
        fprintf(stderr,"reswrap: error reading resource file: \"%s\"\n",argv[arg]);
        break;
        }

      /* Reset for next time */
      needepilogue=1;
      resource=0;
      }
    }

  /* Output epilogue */
  if(needepilogue){
    epilogue(&opts);
    }

  return 0;
  }
